lib/quicly.c

```diff
@@ -4375,15 +4681,184 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
     return 0;
 }
 
+static struct st_quicly_spare_cid_t *pick_spare_cid(quicly_conn_t *conn)
+{
+    struct st_quicly_spare_cid_t *cid = NULL;
+    /* find the CID with smallest sequence number */
+    for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
+        struct st_quicly_spare_cid_t *c = &conn->super.peer.spare_cids[i];
+        if (!c->is_active)
+            continue;
+
+        if (cid == NULL || cid->sequence > c->sequence) {
+            cid = c;
+        }
+    }
+
+    return cid;
+}
+
+/**
+ * compares CIDs, returns zero if identical
+ */
+static int cmp_cid(const quicly_cid_t *l, ptls_iovec_t r)
+{
+    if (l->len != r.len)
+        return 1;
+    return memcmp(l->cid, r.base, l->len);
+}
+
 static int handle_new_connection_id_frame(quicly_conn_t *conn, struct st_quicly_handle_payload_state_t *state)
 {
+    int ret;
+    int need_to_replace_cid = 0;
     quicly_new_connection_id_frame_t frame;
-    return quicly_decode_new_connection_id_frame(&state->src, state->end, &frame);
+
+    /* TODO: return error when using zero-length CID */
+
+    if ((ret = quicly_decode_new_connection_id_frame(&state->src, state->end, &frame)) != 0)
+        return ret;
+
+    QUICLY_PROBE(NEW_CONNECTION_ID_RECEIVE, conn, probe_now(), frame.sequence, frame.retire_prior_to);
+
+    if (frame.sequence < conn->super.peer.max_retire_prior_to) {
+        /* An endpoint that receives a NEW_CONNECTION_ID frame with a sequence number smaller than the Retire Prior To
+         * field of a previously received NEW_CONNECTION_ID frame MUST send a corresponding RETIRE_CONNECTION_ID frame
+         * that retires the newly received connection ID, unless it has already done so for that sequence number. (19.15)
+         * TODO: "unless ..." part may not be properly addressed here (we may already have sent the RCID frame for this
+         * sequence) */
+        if (schedule_retire_connection_id(conn, frame.sequence) != 0)
+            return QUICLY_TRANSPORT_ERROR_INTERNAL;
```

Comment:
What does the return code of `schedule_retire_connection_id` mean? If it is an error code, then I think we should be returning the value returned by the function.

As an alternative, I think we can change the return type of `schedule_retire_connection_id` to void?

Reply:
Good catch. `schedule_retire_connection_id` used to fail (had a malloc) but now does not. Changed the function type to `void` to simplify the code.